File: java/leetcode/PolishNotation.java
//Leetcode problem 150 Evaluate Reverse Polish Notation
//Solution written by Xuqiang Fang on 
import java.util.Stack;
class Solution{
    public int evalRPN(String[] tokens){
        if(tokens == null || tokens.length == 0)
            return -1;
        Stack<Integer> stack = new Stack<>(); 
        int a = 0;
        int b = 0;
        try{
            for(String s : tokens){
                //System.out.println("a = " + a);
                //System.out.println("b = " + b);
                switch(s){
                    case "+":
                        a = stack.pop();
                        b = stack.pop();
                        stack.push(a + b);
                        break;
                    case "-":
                        a = stack.pop();
                        b = stack.pop();
                        stack.push(b - a);
                        break;
                    case "*":
                        a = stack.pop();
                        b = stack.pop();
                        stack.push(a * b);
                        break;
                    case "/":
                        a = stack.pop();
                        b = stack.pop();
                        stack.push(b / a);
                        break;
                    default:
                        stack.push(Integer.valueOf(s));
                }
            }
        }
        catch (Exception e){}
        return stack.pop();
    }	
}

public class PolishNotation{
	public static void main(String[] args){
		Solution s = new Solution();
        String[] tokens = {"-2", "1", "+", "3", "*"};
        String[] tokens2 = {"4", "13", "5", "/", "+"};
        System.out.println(s.evalRPN(tokens));
        System.out.println(s.evalRPN(tokens2));
	}
}

File: cpp/leetcode/PolishNotation.cpp
//Leetcode Problem No 150 Evaluate Reverse Polish Notation
//Solution written by Xuqiang Fang on 23 Oct, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class Solution{
public:
    int evalRPN(vector<string>& t) {
        if(t.size() == 0) return 0;
        stack<int> s;
        int a = 0;
        int b = 0;
        for(int i=0; i<t.size(); ++i){
            if(!isdigit(t[i][t[i].length()-1])){
                a = s.top();
                s.pop();
                b = s.top();
                s.pop();
                if(t[i] == "+"){
                    s.push(a+b);
                }
                else if(t[i] == "-"){
                    s.push(b-a);
                }
                else if(t[i] == "*"){
                    s.push(a*b);
                }
                else if(t[i] == "/"){
                    s.push(b/a);
                }
            }
            else{
                s.push(stoi(t[i]));
            }
        }
        return s.top();
    }
};

int main(){
    Solution s;
    vector<string> t{"2","1","+","3","*"};
    cout << s.evalRPN(t) << endl;
    t = {"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"};
    cout << s.evalRPN(t) << endl;
    return 0;
}
